### PR TITLE
[SPARK-42783][SQL] Infer window group limit should run as late as possible

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
https://github.com/apache/spark/pull/40142 have an unrelated change and is actually a regression. The change let infer window group limit runs early.

Infer window group limit should run as late as possible, it is more safe.

### Why are the changes needed?
Infer window group limit should run as late as possible.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
Exists test cases.

Manually generate the micro benchmark.

```
Benchmark Top-K:                                                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------
ROW_NUMBER (PARTITION: , WindowGroupLimit: false)                        10972          11739         765          1.9         523.2       1.0X
ROW_NUMBER (PARTITION: , WindowGroupLimit: true)                          1700           1738          29         12.3          81.0       6.5X
ROW_NUMBER (PARTITION: PARTITION BY b, WindowGroupLimit: false)          24317          24452         113          0.9        1159.5       0.5X
ROW_NUMBER (PARTITION: PARTITION BY b, WindowGroupLimit: true)            6608           6965         348          3.2         315.1       1.7X
RANK (PARTITION: , WindowGroupLimit: false)                              11549          11850         160          1.8         550.7       1.0X
RANK (PARTITION: , WindowGroupLimit: true)                                2916           3211         267          7.2         139.1       3.8X
RANK (PARTITION: PARTITION BY b, WindowGroupLimit: false)                24736          25951         565          0.8        1179.5       0.4X
RANK (PARTITION: PARTITION BY b, WindowGroupLimit: true)                  6825           7256         497          3.1         325.5       1.6X
DENSE_RANK (PARTITION: , WindowGroupLimit: false)                        11857          12513         652          1.8         565.4       0.9X
DENSE_RANK (PARTITION: , WindowGroupLimit: true)                          2721           2937         113          7.7         129.8       4.0X
DENSE_RANK (PARTITION: PARTITION BY b, WindowGroupLimit: false)          24976          25686         760          0.8        1191.0       0.4X
DENSE_RANK (PARTITION: PARTITION BY b, WindowGroupLimit: true)            6568           6884         364          3.2         313.2       1.7X

```

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala [hunk 1]
```diff
@@ -84,6 +84,11 @@ class SparkOptimizer(
       PushPredicateThroughNonJoin,
       PushProjectionThroughLimit,
       RemoveNoopOperators) :+
+    Batch("Infer window group limit", Once,
+      InferWindowGroupLimit,
+      LimitPushDown,
+      LimitPushDownThroughWindow,
+      EliminateLimits) :+
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala:91]: May I ask why this is different from #40142 ? At that time, we only remove one rule. So, I guess we will recover only `InferWindowGroupLimit` here.
```
- Batch("Insert window group limit", Once, InsertWindowGroupLimit) :+
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala [hunk 2]
```diff
@@ -84,11 +84,11 @@ object InferWindowGroupLimit extends Rule[LogicalPlan] with PredicateHelper {
           }
           // Pick a rank-like function with the smallest limit
           selectedLimits.minBy(_._1) match {
-            case (limit, rankLikeFunction) if limit <= conf.windowGroupLimitThreshold =>
+            case (limit, rankLikeFunction) if limit <= conf.windowGroupLimitThreshold &&
+              child.maxRows.forall(_ > limit) =>
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala:88]: Is this inevitable when we move the rule?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala (hunk 1 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala:91]: https://github.com/apache/spark/pull/40142 just extracts the `Limit` from `Window`. But we still need other rules about `Limit` for further optimization. e.g. `LimitPushDownThroughWindow`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala:88]: This is to ensure the rule's idempotence.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala:88]: So, before this reverting, the idempotence was ensured?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala:88]: In previous, the idempotence not be exposed.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala:88]: I'm also curious about why we suddenly detect idempotence issue. The batch is still Once.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InferWindowGroupLimit.scala:88]: In previous, the batch only contains `InferWindowGroupLimit`. After extracts one Limit between Filter and Window, so the plan not meet `InferWindowGroupLimit` again.

This PR added other rules into the same batch, so the Limit extracted by `InferWindowGroupLimit`, will be pushed down to Window. So the plan still meet `InferWindowGroupLimit`.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan @dongjoon-hyun Thank you !

**Pull request closed without merging**